### PR TITLE
feat: compact_noop_update_after_outer_join

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

to solve https://github.com/risingwavelabs/risingwave/issues/17450

use the SQL to test if is takes effect.
```SQL
create table t1(v int, k int primary key);
create table d(vv int, k int primary key);
INSERT into t1 VALUES (1,1);
INSERT into d values (1,1);

create materialized view mv as 
select t1.v, t1.k 
from t1 
full outer join d d1 using(k)
full outer join d d2 using(k)
full outer join d d3 using(k)
full outer join d d4 using(k)
full outer join d d5 using(k)
full outer join d d6 using(k)
full outer join d d7 using(k)
full outer join d d8 using(k)
full outer join d d9 using(k)
full outer join d d10 using(k);

UPDATE t1 set v = 2; 
```
before
image
after
image

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by fuyufjh** [comment]
When `need_compact` is true, the overhead will be buffering all output data in memory. Considering that we have introduced MemTable spilling to avoid OOM of Join, this overhead will bring OOM back. 

Also, the side effect of making `U-/U+` into `-/+` also looks bad to me.

**Comment by st1page** [comment]
> When `need_compact` is true, the overhead will be buffering all output data in memory. Considering that we have introduced MemTable spilling to avoid OOM of Join, this overhead will bring OOM back.

This PR will not buffer the chunks, instead, it will only do compact in each chunk.
 
> Also, the side effect of making `U-/U+` into `-/+` also looks bad to me.

I still think the  `U-/U+` here is not dependable because it does not take into account the stream key, which can easily lead to different stream keys for a pair of U- and U+. This has already caused some bugs in the previous implementation of the over window. https://github.com/risingwavelabs/risingwave/issues/12539

**Comment by fuyufjh** [comment]
> > When `need_compact` is true, the overhead will be buffering all output data in memory. Considering that we have introduced MemTable spilling to avoid OOM of Join, this overhead will bring OOM back.
> 
> This PR will not buffer the chunks, instead, it will only do compact in each chunk.

Hmmm, I think the problem #17450 never happen in one chunk. It's shown in your example: https://github.com/risingwavelabs/risingwave/pull/17449/files

**Comment by st1page** [comment]
> > > When `need_compact` is true, the overhead will be buffering all output data in memory. Considering that we have introduced MemTable spilling to avoid OOM of Join, this overhead will bring OOM back.
> > 
> > 
> > This PR will not buffer the chunks, instead, it will only do compact in each chunk.
> 
> Hmmm, I think the problem #17450 never happen in one chunk. It's shown in your example: https://github.com/risingwavelabs/risingwave/pull/17449/files

I am just realize that we can just compact the output chunk to solve the issue. 
```
                U- 1 1 1 1
                U+ . . 1 1
                U- . . 1 1
                U+ 1 2 1 1
```
will be 
```
                - 1 1 1 1
                + 1 2 1 1
```

File: src/stream/src/executor/join/builder.rs [hunk 1]
```diff
@@ -156,6 +157,29 @@ impl<const T: JoinTypePrimitive, const SIDE: SideTypePrimitive> JoinChunkBuilder
...
+                c.set_vis(i - 1, false);
+                c.set_vis(i, false);
+                c.set_op(i + 1, Op::Insert);
+                i += 3;
+            } else {
+                i += 1;
+            }
+        }
+        c.into()
+    }
```
Comment (st1page) [inline src/stream/src/executor/join/builder.rs:181]: A less intrusive and lower overhead approach, PTAL again when you have time @fuyufjh .

File: src/stream/src/executor/join/builder.rs [hunk 2]
```diff
@@ -156,6 +157,29 @@ impl<const T: JoinTypePrimitive, const SIDE: SideTypePrimitive> JoinChunkBuilder
...
 
+    pub fn post_process(c: StreamChunk) -> StreamChunk {
+        let mut c = StreamChunkMut::from(c);
+
+        // NOTE(st1page): remove the pattern `UpdateDel(k, old), UpdateIns(k, NULL), UpdateDel(k, NULL),  UpdateIns(k, new)`
+        // to avoid this issue <https://github.com/risingwavelabs/risingwave/issues/17450>
+        let mut i = 1;
+        while i < c.capacity() {
+            if c.op(i - 1) == Op::UpdateInsert
+                && c.op(i) == Op::UpdateDelete
```
Comment (fuyufjh) [inline src/stream/src/executor/join/builder.rs:168]: LGTM, but I recommend adding an assertion (or check) for row i-2 and i+1 e.g.

```suggestion
            if c.op(i - 2) == Op::UpdateDelete
                && c.op(i - 1) == Op::UpdateInsert
                && c.op(i) == Op::UpdateDelete
                && c.op(i + 1) == Op::UpdateInsert
```

**Review by fuyufjh — APPROVED** [review]

File: src/stream/src/executor/join/builder.rs [hunk 3]
```diff
@@ -156,6 +157,29 @@ impl<const T: JoinTypePrimitive, const SIDE: SideTypePrimitive> JoinChunkBuilder
         }
     }
 
+    pub fn post_process(c: StreamChunk) -> StreamChunk {
+        let mut c = StreamChunkMut::from(c);
```
Comment (yuhao-su) [inline src/stream/src/executor/join/builder.rs:161]: Please add some comment about this.

Also is it possible to add this in `append_row_matched` etc.

Comment (st1page) [reply src/stream/src/executor/join/builder.rs:161]: only do it on the `JoinChunkBuilder` but not for the `JoinStreamChunkBuilder`

Comment (xxchan) [reply src/stream/src/executor/join/builder.rs:161]: Why not adding comments :(

Comment (st1page) [reply src/stream/src/executor/join/builder.rs:161]: I forgot. 😶‍🌫️

**Pull request merged**